chain/client/src/chunk_executor_actor.rs

```diff
@@ -0,0 +1,516 @@
+    /// Lock the value of mutable validator signer for the duration of a request to ensure consistency.
+    /// Please note that the locked value should not be stored anywhere or passed through the thread boundary.
+    validator_signer: MutableValidatorSigner,
+    shard_tracker: ShardTracker,
+    network_adapter: PeerManagerAdapter,
+
+    /// Receipts originating from block keyed by block hash.
+    block_receipts_cache: LruCache<CryptoHash, HashMap<(ShardId, ShardId), ReceiptProof>>,
+    /// Next block hashes keyed by block hash.
+    next_block_hashes: LruCache<CryptoHash, Vec<CryptoHash>>,
+
+    // Hash of the genesis block.
+    genesis_hash: CryptoHash,
+}
+
+impl ChunkExecutorActor {
+    pub fn new(
+        store: Store,
+        genesis: &ChainGenesis,
+        genesis_hash: CryptoHash,
+        runtime_adapter: Arc<dyn RuntimeAdapter>,
+        epoch_manager: Arc<dyn EpochManagerAdapter>,
+        validator_signer: MutableValidatorSigner,
+        shard_tracker: ShardTracker,
+        network_adapter: PeerManagerAdapter,
+        block_receipts_cache_capacity: NonZeroUsize,
+        next_block_hashes_cache_capacity: NonZeroUsize,
+    ) -> Self {
+        Self {
+            chain_store: ChainStore::new(store, true, genesis.transaction_validity_period),
+            runtime_adapter,
+            epoch_manager,
+            validator_signer,
+            shard_tracker,
+            network_adapter,
+            block_receipts_cache: LruCache::new(block_receipts_cache_capacity),
+            next_block_hashes: LruCache::new(next_block_hashes_cache_capacity),
+            genesis_hash,
+        }
+    }
+}
+
+impl near_async::messaging::Actor for ChunkExecutorActor {}
+
+/// Message with incoming receipts corresponding to the block.
+/// Eventually this would be handled properly with data availability layer.
+/// For now this is useful to do testing with test loop.
+#[derive(actix::Message, Debug)]
+#[rtype(result = "()")]
+pub struct ExecutorIncomingReceipts {
+    pub block_hash: CryptoHash,
+    pub receipt_proofs: Vec<ReceiptProof>,
+}
+
+/// Message that should be sent once block is processed to indicate that it's available for
+/// execution.
+#[derive(actix::Message, Debug)]
+#[rtype(result = "()")]
+pub struct ExecutorBlock {
+    pub block_hash: CryptoHash,
+}
+
+impl Handler<ExecutorIncomingReceipts> for ChunkExecutorActor {
+    fn handle(
+        &mut self,
+        ExecutorIncomingReceipts { block_hash, receipt_proofs }: ExecutorIncomingReceipts,
+    ) {
+        let block_receipts =
+            self.block_receipts_cache.get_or_insert_mut(block_hash, || HashMap::new());
+        for proof in receipt_proofs {
+            block_receipts.insert((proof.1.from_shard_id, proof.1.to_shard_id), proof);
+        }
+
+        let me = self.validator_signer.get().map(|signer| signer.validator_id().clone());
+
+        let Some(next_block_hashes) = self.next_block_hashes.pop(&block_hash) else {
```

Comment:
did you mean to use `pop` here 🤔 ?
```
    /// Removes and returns the value corresponding to the key from the cache or
    /// `None` if it does not exist.
```
that doesn't seem right since the block will be removed after receiving the first set of receipts which is not enough to apply chunks

Reply:
Good catch! I have no idea why I though that we would have all incoming receipts by this time.
Fixed in [2f45d21](https://github.com/near/nearcore/pull/13637/commits/2f45d214c8af1b9906ecbe1fc8bd266766d4a3f2)